src/modules/silentpayments/main_impl.h

```diff
@@ -322,4 +322,308 @@ int secp256k1_silentpayments_recipient_create_labelled_spend_pubkey(const secp25
+    return ret;
+}
+
+int secp256k1_silentpayments_recipient_public_data_parse(const secp256k1_context *ctx, secp256k1_silentpayments_public_data *public_data, const unsigned char *input33) {
+    size_t inputlen = 33;
+    size_t pubkeylen = 65;
+    secp256k1_pubkey pubkey;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(input33 != NULL);
+    /* Since an attacker can send us malicious data that looks like a serialized public key but is not, fail early */
+    if (!secp256k1_ec_pubkey_parse(ctx, &pubkey, input33, inputlen)) {
+        return 0;
+    }
+    public_data->data[0] = 1;
+    secp256k1_ec_pubkey_serialize(ctx, &public_data->data[1], &pubkeylen, &pubkey, SECP256K1_EC_UNCOMPRESSED);
+    memset(&public_data->data[1 + pubkeylen], 0, 32);
+    return 1;
+}
+
+int secp256k1_silentpayments_recipient_scan_outputs(
+    const secp256k1_context *ctx,
+    secp256k1_silentpayments_found_output **found_outputs, size_t *n_found_outputs,
+    const secp256k1_xonly_pubkey * const *tx_outputs, size_t n_tx_outputs,
+    const unsigned char *recipient_scan_key,
+    const secp256k1_silentpayments_public_data *public_data,
+    const secp256k1_pubkey *recipient_spend_pubkey,
+    const secp256k1_silentpayments_label_lookup label_lookup,
+    const void *label_context
+) {
+    secp256k1_scalar t_k_scalar, rsk_scalar;
+    secp256k1_ge label_ge, recipient_spend_pubkey_ge, A_sum_ge;
+    secp256k1_pubkey A_sum;
+    secp256k1_xonly_pubkey P_output_xonly;
+    unsigned char shared_secret[33];
+    const unsigned char *label_tweak = NULL;
+    size_t i, k, n_found, found_idx;
+    int found, combined;
+    int ret = 1;
+
+    /* Sanity check inputs */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(found_outputs != NULL);
+    ARG_CHECK(n_found_outputs != NULL);
+    ARG_CHECK(tx_outputs != NULL);
+    ARG_CHECK(n_tx_outputs > 0);
+    ARG_CHECK(recipient_scan_key != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(recipient_spend_pubkey != NULL);
+    if (label_lookup != NULL) {
+        ARG_CHECK(label_context != NULL);
+    } else {
+        ARG_CHECK(label_context == NULL);
+    }
+    /* TODO: do we need a _cmov call here to avoid leaking information about the scan key?
+     * Recall: a scan key isnt really "secret" data in that leaking the scan key will only leak privacy
+     * In this respect, a scan key is functionally equivalent to an xpub
+     */
+    ret &= secp256k1_scalar_set_b32_seckey(&rsk_scalar, recipient_scan_key);
+    ret &= secp256k1_silentpayments_recipient_public_data_load_pubkey(ctx, &A_sum, public_data);
+    ret &= secp256k1_pubkey_load(ctx, &A_sum_ge, &A_sum);
+    ret &= secp256k1_pubkey_load(ctx, &recipient_spend_pubkey_ge, recipient_spend_pubkey);
+    /* If there is something wrong with the recipient scan key, recipient spend pubkey, or the public data, return early */
+    if (!ret) {
+        return 0;
+    }
+    combined = (int)public_data->data[0];
+    if (!combined) {
```

Comment:
https://github.com/bitcoin-core/secp256k1/pull/1519/commits/94c6e1f54299dfc83aa34bc92f9f861808449a23: There's something I don't understand here. Why is `secp256k1_silentpayments_public_data` 98 bytes long? If we only kept the "combined" format `input_hash . A_sum` like we do in `secp256k1_silentpayments_recipient_public_data_serialize`, we would not need this `if  (!combined)` block. What are the merits of keeping `A_sum || input_hash` in public data instead of `input_hash . A_sum`?

Reply:
Thanks @stratospher. I understand that we have to check if combined is set so we know how to handle the public data and I get that light clients need `A_sum . Input_hash`, but what I'm confused about is why don't we do `A_sum . Input_hash` in `secp256k1_silentpayments_recipient_public_data_create` so we only need 32 bytes for public data and we don't need the `secp256k1_silentpayments_recipient_public_data_serialize` function because the data is already presentable to light clients? I'm guessing there's some reason to not combine the `A_sum` and `input_hash` at the beginning?